### PR TITLE
ui grid in complex voucher

### LINKED ISSUE #786: (feature) Complex Vouchers should load Cash Payments
The Complex Vouchers page allows a user to import records from Cash Payments, Patient Invoices, and previous Vouchers.  Currently, only the Patient Invoices and Voucher grids are defined, but not the Cash Payments grid.  See below:
complexvouchercashpaymentsmodal
_Fig 1: Empty Cash Payments Grid_

Before 2.x can be marked as complete, this grid must be finished, with the ability to import data from the cash payments table.

### PR DESCRIPTION
**Pull request by mbayopanda**
This PR : 
- Replace `html` table by a `ui-grid`
- Add payment reference and fixing
- Add a search button in the reference modal for finding references 
- Replace the `typeahead` by a `ui-select` in entity modal 
- Fix grid resizing
- Fix transaction reference 

close #786

### CODE REVIEW

File: client/src/partials/templates/modals/findEntity.modal.html [hunk 1]
```diff
@@ -38,14 +38,20 @@
         </div>
         <!-- /dropdown button -->
 
-        <input
-          type="text"
-          class="form-control"
+        <ui-select
```
Comment (jniles) [inline client/src/partials/templates/modals/findEntity.modal.html:41]: :+1:  Nice improvement.

File: client/src/partials/templates/modals/findReference.modal.js [hunk 2]
```diff
@@ -81,11 +94,22 @@ function FindReferenceModalController(Instance, Debtor, Creditor, Voucher, Filte
...
-    /**
-    * @fixme the balance value is not correct for the document references
-    * The correct value is the amount of the invoice in the receipt
-    */
-    return;
+    Cash.read()
+    .then(function (list) {
+      var costTemplate =
+        '<div class="ui-grid-cell-contents text-right">' +
+          '{{ row.entity.amount | currency: grid.appScope.enterprise.currency_id }}' +
```
Comment (jniles) [inline client/src/partials/templates/modals/findReference.modal.js:101]: :+1:

File: client/src/partials/vouchers/complex.html [hunk 3]
```diff
@@ -101,164 +101,49 @@
...
 
-              <span ng-class="{'text-danger': !ComplexVoucherCtrl.validInput, 'text-success': ComplexVoucherCtrl.validInput}" ng-hide="ComplexVoucherCtrl.posted">
-                <span ng-class="ComplexVoucherCtrl.notifyMessage.icon"></span>
-                {{ ComplexVoucherCtrl.notifyMessage.label | translate }}
-              </span>
+          <!-- the grid  -->
+          <div id="voucherGridId"
+            ui-grid="ComplexVoucherCtrl.gridOptions"
+            ui-grid-auto-resize
+            ui-grid-resize-columns>
```
Comment (jniles) [inline client/src/partials/vouchers/complex.html]: I don't think resizing columns is a good idea on a grid with inputs.  What do you think?  Is it really necessary?

**Review by jniles — CHANGES_REQUESTED** [review]
@mbayopanda, this looks pretty good.

I'm concerned about putting a ui-select in the `entity` modal.  If we have thousands of patients, this may become a performance bottleneck (by default, all patients will be rendered as debtors in the dropdown.  While `ui-select` is a very useful and powerful tool, I don't think that's a good use case.

**Comment by mbayopanda** [comment]
@jniles could you review this PR

File: test/end-to-end/shared/FormUtils.js [hunk 4]
```diff
@@ -167,6 +167,33 @@ module.exports = {
...
+   * @description
+   * Selects an option from the ui-select dropdown
+   * which have the `append-to-body` option set to true
+   *
+   * @param {String} model - the ng-model target to select
+   * @param {String} label - the text of the option element to choose
+   * @param {Element} anchor - a protractor element to search within
+   * @returns {Element} - a protractor option element
+   */
+  uiSelectAppended: function uiSelectAppended(model, label, anchor) {
```
Comment (jniles) [inline test/end-to-end/shared/FormUtils.js:181]: Cool!

File: test/end-to-end/shared/FormUtils.js [hunk 5]
```diff
@@ -167,6 +167,33 @@ module.exports = {
...
+    const externalAnchor = $('body > div.ui-select-bootstrap.dropdown');
+
+    // click on the element
+    anchor = anchor || $('body');
```
Comment (jniles) [inline test/end-to-end/shared/FormUtils.js]: I don't think you ever use the anchor.  Could it be removed?

**Review by jniles — APPROVED** [review]
LGTM

**Pull request merged**